crates/network-scanner-proto/src/icmp_v4.rs

```diff
@@ -0,0 +1,400 @@
+                sequence: BigEndian::read_u16(&bytes[6..8]),
+            },
+            13 => {
+                if packet_len < 20 {
+                    return Err(PacketParseError::PacketTooSmall(bytes.len()));
+                }
+                Icmpv4Message::Timestamp {
+                    identifier: BigEndian::read_u16(&bytes[4..6]),
+                    sequence: BigEndian::read_u16(&bytes[6..8]),
+                    originate: BigEndian::read_u32(&bytes[8..12]),
+                    receive: BigEndian::read_u32(&bytes[12..16]),
+                    transmit: BigEndian::read_u32(&bytes[16..20]),
+                }
+            }
+            14 => {
+                if packet_len < 20 {
+                    return Err(PacketParseError::PacketTooSmall(bytes.len()));
+                }
+                Icmpv4Message::TimestampReply {
+                    identifier: BigEndian::read_u16(&bytes[4..6]),
+                    sequence: BigEndian::read_u16(&bytes[6..8]),
+                    originate: BigEndian::read_u32(&bytes[8..12]),
+                    receive: BigEndian::read_u32(&bytes[12..16]),
+                    transmit: BigEndian::read_u32(&bytes[16..20]),
+                }
+            }
+            t => {
+                dbg!(bytes);
+                return Err(PacketParseError::UnrecognizedICMPType(t));
+            }
+        };
+        return Ok(Icmpv4Packet {
+            code: code,
+            checksum: checksum,
+            message: message,
+        });
+    }
+
+    /// Get this packet serialized to bytes suitable for sending on the wire.
+    pub fn to_bytes(&self, with_checksum: bool) -> Vec<u8> {
+        let mut bytes = Vec::new();
+        bytes.push(self.message.get_type() as u8);
+        bytes.push(self.code);
+        let mut buf = vec![0; 2];
+        BigEndian::write_u16(&mut buf, if with_checksum { self.checksum } else { 0 });
+        bytes.append(&mut buf);
+        bytes.append(&mut self.message.get_bytes());
+        return bytes;
+    }
+
+    pub fn calculate_checksum(&self) -> u16 {
+        let mut sum = 0u32;
+        let bytes = self.to_bytes(false);
+        sum += sum_big_endian_words(&bytes);
+
+        while sum >> 16 != 0 {
+            sum = (sum >> 16) + (sum & 0xFFFF);
+        }
+        !sum as u16
+    }
+
+    /// Populate the checksum field of this Packet.
+    pub fn with_checksum(mut self) -> Self {
+        self.checksum = self.calculate_checksum();
+        self
+    }
+
+    pub fn from_message(message: Icmpv4Message) -> Self {
+        Self {
+            code: 0,
+            checksum: 0,
+            message: message,
+        }
+        .with_checksum()
+    }
+}
+
+/// Error type returned by parsing the ICMP packets.
+#[derive(Debug)]
+pub enum PacketParseError {
+    /// Not enough bytes to properly parse the packet from.
+    PacketTooSmall(usize),
+    /// An unrecognized ICMP type.
+    UnrecognizedICMPType(u8),
+}
```

Comment:
suggestion: You could also implement the `std::error::Error` trait. Side bonus: `anyhow` is working better with error types implementing it.